File: python/leanfe/leanfe.py
"""
Fast high-dimensional fixed effects regression.

Main entry point for the package with support for Polars and DuckDB backends.
"""
from leanfe.result import LeanFEResult
from typing import Literal
import polars as pl
from duckdb import DuckDBPyConnection
from leanfe.polars_impl import leanfe_polars
from leanfe.duckdb_impl import leanfe_duckdb


def leanfe(
    data: str | pl.DataFrame | pl.LazyFrame | None = None,
    y_col: str | None = None,
    x_cols: list[str] | None = None,
    fe_cols: list[str]  = [],
    formula: str | None = None,
    strategy: str = 'auto',
    weights: str | None = None,
    demean_tol: float = 1e-10,
    max_iter: int = 500,
    vcov: str = "iid",
    cluster_cols: list[str] | None = None,
    ssc: bool = True,
    sample_frac: float | None = None,
    backend: Literal["polars", "duckdb"] = "polars",
    con: DuckDBPyConnection | None = None
) -> LeanFEResult:
    """
    Fast fixed effects regression using Polars or DuckDB backend.
    
    This is the main entry point for the package. By default uses the Polars
    backend for speed; use DuckDB for very large datasets that don't fit in memory.
    
    Parameters
    ----------
    data : str or polars.DataFrame
        Input data: either a Polars DataFrame or path to a Parquet file.
    y_col : str, optional
        Dependent variable column name (optional if formula provided).
    x_cols : list of str, optional
        Independent variable column names (optional if formula provided).
    fe_cols : list of str, optional
        Fixed effect column names (optional if formula provided).
    formula : str, optional
        R-style formula: "y ~ x1 + x2 + x:i(factor) | fe1 + fe2 | z1 + z2" (IV).
        Supports:
        - Regular variables: x1, x2
        - Factor variables: i(region) → automatic dummy expansion
        - Interactions: treatment:i(region) → heterogeneous effects
        - Instruments (IV): third part after second |
    weights : str, optional
        Column name for regression weights (WLS).
    demean_tol : float, default 1e-5
        Convergence tolerance for iterative demeaning.
    max_iter : int, default 500
        Maximum iterations for demeaning.
    vcov : str, default "iid"
        Variance-covariance estimator: "iid", "HC1", or "cluster".
    cluster_cols : list of str, optional
        Clustering variables (required if vcov="cluster").
    ssc : bool, default False
        Small sample correction for clustered standard errors.
    sample_frac : float, optional
        Fraction of data to sample (e.g., 0.1 for 10%).
    backend : {"polars", "duckdb"}, default "polars"
        Computation backend to use:
        - "polars": Faster, better for data that fits in memory (default)
        - "duckdb": Lower memory usage, better for very large datasets
    
    Returns
    -------
    dict
        Dictionary containing:
        - coefs: dict mapping variable names to coefficient estimates
        - std_errors: dict mapping variable names to standard errors
        - n_obs: number of observations used
        - iterations: number of demeaning iterations
        - vcov_type: type of variance estimator used
        - is_iv: whether IV/2SLS was used
        - n_instruments: number of instruments (if IV)
        - n_clusters: number of clusters (if clustered SEs)
    
    Notes
    -----
    **Choosing a Backend:**
    
    Use Polars (default) when:
    - Speed is the priority
    - Data fits comfortably in memory
    - You need the fastest possible execution
    
    Use DuckDB when:
    - Dataset is larger than available RAM
    - Memory is constrained
    - Reading directly from parquet files without loading into memory
    
    **Performance Comparison (12.7M obs, 4 FEs):**
    - Polars: ~6s, ~300 MB memory
    - DuckDB: ~18s, ~30 MB memory
    
    Examples
    --------
    Basic usage with formula:
    
    >>> from leanfe import leanfe
    >>> result = leanfe(df, formula="y ~ treatment | customer + product")
    >>> print(result['coefs']['treatment'])
    
    With clustered standard errors:
    
    >>> result = leanfe(
    ...     df,
    ...     formula="y ~ treatment | customer + product",
    ...     vcov="cluster",
    ...     cluster_cols=["customer"]
    ... )
    
    Using DuckDB for large datasets:
    
    >>> result = leanfe(
    ...     "large_data.parquet",
    ...     formula="y ~ treatment | fe1 + fe2",
    ...     backend="duckdb"
    ... )
    
    Difference-in-Differences:
    
    >>> result = leanfe(
    ...     df,
    ...     formula="y ~ treated_post | state + year",
    ...     vcov="cluster",
    ...     cluster_cols=["state"]
    ... )
    """
    if backend == "polars":
        con = None
        if data is None:
            raise ValueError(
                f'A dataset must be provided when using backend = {backend}.\n'
                "Example: leanfe(data = pl.scan_parquet('data.pq').collect(), backend='polars')"
            )
        return leanfe_polars(
            data = data,
            y_col = y_col,
            x_cols = x_cols,
            fe_cols = fe_cols,
            formula = formula,
            strategy = strategy,
            weights = weights,
            demean_tol = demean_tol,
            max_iter = max_iter,
            vcov = vcov,
            cluster_cols = cluster_cols,
            ssc = ssc,
            sample_frac = sample_frac
        )
    elif backend == "duckdb":
        if con is None:
            raise ValueError(
                f'A DuckDBPyConnection must be provided when using backend = {backend}.\n'
                "Example: leanfe(..., backend='duckdb', con=duckdb.connect())\n"
                "Users may want to set hard limits for max_threads and memory_limit in duckdb.con()"
            )
        return leanfe_duckdb(
            data = data,
            y_col = y_col,
            x_cols = x_cols,
            fe_cols = fe_cols,
            formula = formula,
            strategy = strategy,
            weights = weights,
            demean_tol = demean_tol,
            max_iter = max_iter,
            vcov = vcov,
            cluster_cols = cluster_cols,
            ssc = ssc,
            sample_frac = sample_frac,
            con = con
        )
    else:
        raise ValueError(f"backend must be 'polars' or 'duckdb', got '{backend}'")

File: r/R/leanfe.R
#' Fast High-Dimensional Fixed Effects Regression
#'
#' Main entry point for fixed effects regression using either Polars or DuckDB backend.
#' By default uses Polars for speed; use DuckDB for very large datasets.
#'
#' @description
#' This is the recommended entry point for the package. It provides a unified API
#' that dispatches to either the Polars or DuckDB backend based on the `backend` parameter.
#'
#' @param data Polars DataFrame, R data.frame, or path to parquet file.
#' @param y_col Dependent variable column name (optional if formula provided).
#' @param x_cols Character vector of independent variable column names (optional if formula provided).
#' @param fe_cols Character vector of fixed effect column names (optional if formula provided).
#' @param formula R-style formula: "y ~ x1 + x2 | fe1 + fe2" or "y ~ x | fe | z" (IV).
#'   Supports:
#'   \itemize{
#'     \item Regular variables: x1, x2
#'     \item Factor variables: i(region) - automatic dummy expansion
#'     \item Interactions: treatment:i(region) - heterogeneous effects
#'     \item Instruments (IV): third part after second |
#'   }
#' @param weights Column name for regression weights (WLS).
#' @param demean_tol Convergence tolerance for iterative demeaning (default: 1e-5).
#' @param max_iter Maximum iterations for demeaning (default: 500).
#' @param vcov Variance-covariance estimator: "iid", "HC1", or "cluster" (default: "iid").
#' @param cluster_cols Character vector of clustering variables (required if vcov="cluster").
#' @param ssc Logical, small sample correction for clustered SEs (default: FALSE).
#' @param sample_frac Fraction of data to sample, e.g., 0.1 for 10% (default: NULL).
#' @param backend Computation backend: "polars" (default) or "duckdb".
#'   \itemize{
#'     \item "polars": Faster, better for data that fits in memory
#'     \item "duckdb": Lower memory usage, better for very large datasets
#'   }
#'
#' @return A list containing:
#' \describe{
#'   \item{coefficients}{Named vector of coefficient estimates}
#'   \item{std_errors}{Named vector of standard errors}
#'   \item{n_obs}{Number of observations used}
#'   \item{iterations}{Number of demeaning iterations}
#'   \item{vcov_type}{Type of variance estimator used}
#'   \item{n_clusters}{Number of clusters (if clustered SEs)}
#' }
#'
#' @details
#' \strong{Choosing a Backend:}
#'
#' Use Polars (default) when:
#' \itemize{
#'   \item Speed is the priority
#'   \item Data fits comfortably in memory
#'   \item You need the fastest possible execution
#' }
#'
#' Use DuckDB when:
#' \itemize{
#'   \item Dataset is larger than available RAM
#'   \item Memory is constrained
#'   \item Reading directly from parquet files without loading into memory
#' }
#'
#' \strong{Performance Comparison (12.7M obs, 4 FEs):}
#' \itemize{
#'   \item Polars: ~15s, ~1 GB memory
#'   \item DuckDB: ~18s, ~700 MB memory
#' }
#'
#' @examples
#' \dontrun{
#' library(polars)
#'
#' # Basic usage with formula
#' df <- pl$read_parquet("data.parquet")
#' result <- leanfe(df, formula = "y ~ treatment | customer + product")
#' print(result$coefficients["treatment"])
#'
#' # With clustered standard errors
#' result <- leanfe(
#'   df,
#'   formula = "y ~ treatment | customer + product",
#'   vcov = "cluster",
#'   cluster_cols = c("customer")
#' )
#'
#' # Using DuckDB for large datasets
#' result <- leanfe(
#'   "large_data.parquet",
#'   formula = "y ~ treatment | fe1 + fe2",
#'   backend = "duckdb"
#' )
#'
#' # Difference-in-Differences
#' result <- leanfe(
#'   df,
#'   formula = "y ~ treated_post | state + year",
#'   vcov = "cluster",
#'   cluster_cols = c("state")
#' )
#' }
#'
#' @export
leanfe <- function(
  data,
  y_col = NULL,
  x_cols = NULL,
  fe_cols = NULL,
  formula = NULL,
  weights = NULL,
  demean_tol = 1e-5,
  max_iter = 500,
  vcov = "iid",
  cluster_cols = NULL,
  ssc = FALSE,
  sample_frac = NULL,
  backend = "polars"
) {
  if (!backend %in% c("polars", "duckdb")) {
    stop(sprintf("backend must be 'polars' or 'duckdb', got '%s'", backend))
  }
  
  if (backend == "polars") {
    leanfe_polars(
      data = data,
      y_col = y_col,
      x_cols = x_cols,
      fe_cols = fe_cols,
      formula = formula,
      weights = weights,
      demean_tol = demean_tol,
      max_iter = max_iter,
      vcov = vcov,
      cluster_cols = cluster_cols,
      ssc = ssc,
      sample_frac = sample_frac
    )
  } else {
    leanfe_duckdb(
      data = data,
      y_col = y_col,
      x_cols = x_cols,
      fe_cols = fe_cols,
      formula = formula,
      weights = weights,
      demean_tol = demean_tol,
      max_iter = max_iter,
      vcov = vcov,
      cluster_cols = cluster_cols,
      ssc = ssc,
      sample_frac = sample_frac
    )
  }
}
